receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,424 @@
+func IDSetFromTraces(data ptrace.Traces) (IDSet, error) {
+	ds := map[UniqueIDAttrVal]bool{}
+	rss := data.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		ils := rss.At(i).ScopeSpans()
+		for j := 0; j < ils.Len(); j++ {
+			ss := ils.At(j).Spans()
+			for k := 0; k < ss.Len(); k++ {
+				elem := ss.At(k)
+				key, exists := elem.Attributes().Get(UniqueIDAttrName)
+				if !exists {
+					return ds, fmt.Errorf("invalid data element, attribute %q is missing", UniqueIDAttrName)
+				}
+				if key.Type() != pcommon.ValueTypeStr {
+					return ds, fmt.Errorf("invalid data element, attribute %q is wrong type %v", UniqueIDAttrName, key.Type())
+				}
+				ds[key.Str()] = true
+			}
+		}
+	}
+	return ds, nil
+}
+
+func (m *mockConsumer) ConsumeLogs(_ context.Context, data plog.Logs) error {
+	ids, err := IDSetFromLogs(data)
+	require.NoError(m.t, err)
+	return m.consume(ids)
+}
+
+// IDSetFromLogs computes an IDSet from given plog.Logs. The IDSet will contain ids of all log records.
+func IDSetFromLogs(data plog.Logs) (IDSet, error) {
+	ds := map[UniqueIDAttrVal]bool{}
+	rss := data.ResourceLogs()
+	for i := 0; i < rss.Len(); i++ {
+		ils := rss.At(i).ScopeLogs()
+		for j := 0; j < ils.Len(); j++ {
+			ss := ils.At(j).LogRecords()
+			for k := 0; k < ss.Len(); k++ {
+				elem := ss.At(k)
+				key, exists := elem.Attributes().Get(UniqueIDAttrName)
+				if !exists {
+					return ds, fmt.Errorf("invalid data element, attribute %q is missing", UniqueIDAttrName)
+				}
+				if key.Type() != pcommon.ValueTypeStr {
+					return ds, fmt.Errorf("invalid data element, attribute %q is wrong type %v", UniqueIDAttrName, key.Type())
+				}
+				ds[key.Str()] = true
+			}
+		}
+	}
+	return ds, nil
+}
+
+// TODO: Implement mockConsumer.ConsumeMetrics()
+
+// consume the elements with the specified ids, regardless of the element data type.
+func (m *mockConsumer) consume(ids IDSet) error {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	// Consult with user-defined decision function to decide what to do with the data.
+	if err := m.consumeDecisionFunc(ids); err != nil {
+		// The decision is to return an error to the receiver.
+
+		if consumererror.IsPermanent(err) {
+			// It is a permanent error, which means we need to drop the data.
+			// Remember the ids of dropped elements.
+			duplicates := m.droppedIds.merge(ids)
+			if len(duplicates) > 0 {
+				require.FailNow(m.t, "elements that were dropped previously were sent again: %v", duplicates)
+			}
+		} else {
+			// It is a non-permanent error. Don't add it to the drop list. Remember the number of
+			// failures to print at the end of the test.
+			m.nonPermanentFailures++
+		}
+		// Return the error to the receiver.
+		return err
+	}
+
+	// The decision is a success. Remember the ids of the data in the accepted list.
+	duplicates := m.acceptedIds.merge(ids)
+	if len(duplicates) > 0 {
+		require.FailNow(m.t, "elements that were accepted previously were sent again: %v", duplicates)
```

Comment:
Should it be `require.FailNowf`?

Or we can have instead of the condition:

```
require.Emptyf(m.t, duplicates, "elements that were accepted previously were sent again: %v", duplicates)
```

I believe it'll be the same effect